msvc/winsetup.bat

```diff
@@ -1,35 +1,65 @@
 
-%windir%\system32\WindowsPowerShell\v1.0\powershell.exe -executionpolicy bypass -NonInteractive -File compare-config-files.ps1 %WIN_CONFIG_H% %CONFIG_H% %CONFIGURE_AC% 2>&1
+:: Extract MONO_VERSION from configure.ac.
+for /f "delims=[] tokens=2" %%a in ('findstr /b /c:"AC_INIT(mono, [" %CONFIGURE_AC%') do (
+	set MONO_VERSION=%%a
+)
 
-IF NOT %ERRORLEVEL% == 0 (
-	ECHO copy %WIN_CONFIG_H% %CONFIG_H%
-	copy %WIN_CONFIG_H% %CONFIG_H%
-	%windir%\system32\WindowsPowerShell\v1.0\powershell.exe -NonInteractive -Command "(Get-Content %CONFIG_H%) -replace '#MONO_VERSION#', (Select-String -path %CONFIGURE_AC% -pattern 'AC_INIT\(mono, \[(.*)\]').Matches[0].Groups[1].Value | Set-Content %CONFIG_H%" 2>&1
-	%windir%\system32\WindowsPowerShell\v1.0\powershell.exe -NonInteractive -Command "$mono_version=[int[]](Select-String -path %CONFIGURE_AC% -pattern 'AC_INIT\(mono, \[(.*)\]').Matches[0].Groups[1].Value.Split('.'); $corlib_counter=[int](Select-String -path %CONFIGURE_AC% -pattern 'MONO_CORLIB_COUNTER=(.*)').Matches[0].Groups[1].Value; (Get-Content %CONFIG_H%) -replace '#MONO_CORLIB_VERSION#',('1{0:00}{1:00}{2:00}{3:000}' -f $mono_version[0],$mono_version[1],0,$corlib_counter) | Set-Content %CONFIG_H%" 2>&1
+:: Split MONO_VERSION into three parts.
+for /f "delims=. tokens=1-3" %%a in ('echo %MONO_VERSION%') do (
+	set MONO_VERSION_MAJOR=%%a
+	set MONO_VERSION_MINOR=%%b
+	set MONO_VERSION_PATCH=%%c
 )
+:: configure.ac hardcodes this.
+set MONO_VERSION_PATCH=00
 
-SET VERSION_CONTENT="#define FULL_VERSION \"Visual Studio built mono\""
-%windir%\system32\WindowsPowerShell\v1.0\powershell.exe -executionpolicy bypass -NonInteractive -File compare-config-content.ps1 %VERSION_CONTENT% %VERSION_H%  2>&1
+:: Extract MONO_CORLIB_COUNTER from configure.ac.
+for /f "tokens=*" %%a in ('findstr /b /c:MONO_CORLIB_COUNTER= %CONFIGURE_AC%') do set %%a
 
+:: Pad out version pieces to 2 characters with zeros on left.
+:: corlib_counter same but 3 characters.
+if "%MONO_VERSION_MAJOR:~1%" == "" set MONO_VERSION_MAJOR=0%MONO_VERSION_MAJOR%
+if "%MONO_VERSION_MINOR:~1%" == "" set MONO_VERSION_MINOR=0%MONO_VERSION_MINOR%
+if "%MONO_CORLIB_COUNTER:~2%" == "" set MONO_CORLIB_COUNTER=0%MONO_CORLIB_COUNTER%
+if "%MONO_CORLIB_COUNTER:~2%" == "" set MONO_CORLIB_COUNTER=0%MONO_CORLIB_COUNTER%
 
-IF NOT %ERRORLEVEL% == 0 (
-	ECHO Configure %VERSION_H%
-	ECHO #define FULL_VERSION "Visual Studio built mono"> %VERSION_H%
-)
+set MONO_CORLIB_VERSION=1%MONO_VERSION_MAJOR%%MONO_VERSION_MINOR%%MONO_VERSION_PATCH%%MONO_CORLIB_COUNTER%
 
-ECHO Successfully setup Mono configuration headers.
+:: Remove every define VERSION from config.h and add what we want.
+findstr /v /r /i /c:"#define..*VERSION" %win_config_h% > %monotemp%
+echo #define PACKAGE_VERSION "%MONO_VERSION%" >> %monotemp%
+echo #define VERSION "%MONO_VERSION%" >> %monotemp%
+echo #define MONO_CORLIB_VERSION %MONO_CORLIB_VERSION% >> %monotemp%
```

Comment:
it probably won't matter too much but this places the defines outside of the `#endif` in winconfig.h

Reply:
As long as all the defines are the same, it is ok. You can duplicate defines w/o error or warning.
Since this is Windows-specific file, can also throw in pragma once.

Reply:
It will actually be used by cygwin gcc build when doing make again after build from within Visual Studio, just verified that scenario and the define duplication doesn't generate any additonal warnings or errors, so this should be fine.